File: src/main/java/com/example/sudoku/image/ImagePreprocessor.java
package com.example.sudoku.image;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;

import org.opencv.core.Mat;
import org.opencv.core.MatOfPoint;
import org.opencv.core.MatOfPoint2f;
import org.opencv.core.Point;
import org.opencv.core.Size;
import org.opencv.imgproc.Imgproc;

public class ImagePreprocessor {

    public Mat preprocessSudokuImage(Mat src) {
        Mat gray = new Mat();
        Imgproc.cvtColor(src, gray, Imgproc.COLOR_BGR2GRAY);

        Imgproc.equalizeHist(gray, gray);
        Imgproc.GaussianBlur(gray, gray, new Size(7, 7), 0);

        Mat thresh = new Mat();
        Imgproc.adaptiveThreshold(gray, thresh, 255,
                Imgproc.ADAPTIVE_THRESH_GAUSSIAN_C, Imgproc.THRESH_BINARY_INV, 11, 2);

        Mat kernel = Imgproc.getStructuringElement(Imgproc.MORPH_RECT, new Size(3, 3));
        Imgproc.morphologyEx(thresh, thresh, Imgproc.MORPH_OPEN, kernel);

        List<MatOfPoint> contours = new ArrayList<>();
        Mat hierarchy = new Mat();
        Imgproc.findContours(thresh, contours, hierarchy,
                Imgproc.RETR_EXTERNAL, Imgproc.CHAIN_APPROX_SIMPLE);

        double maxArea = 0;
        MatOfPoint maxContour = null;
        for (MatOfPoint contour : contours) {
            double area = Imgproc.contourArea(contour);
            if (area > maxArea) {
                MatOfPoint2f approxCurve = new MatOfPoint2f();
                MatOfPoint2f contour2f = new MatOfPoint2f(contour.toArray());
                double peri = Imgproc.arcLength(contour2f, true);
                Imgproc.approxPolyDP(contour2f, approxCurve, 0.02 * peri, true);
                if (approxCurve.total() == 4) {
                    maxArea = area;
                    maxContour = new MatOfPoint(approxCurve.toArray());
                }
            }
        }
        if (maxContour == null) throw new RuntimeException("Could not find Sudoku grid");

        Point[] pts = maxContour.toArray();
        Point[] orderedPts = orderPoints(pts);

        double side = 450;
        MatOfPoint2f dest = new MatOfPoint2f(
                new Point(0, 0),
                new Point(side - 1, 0),
                new Point(side - 1, side - 1),
                new Point(0, side - 1));
        MatOfPoint2f srcPts = new MatOfPoint2f(orderedPts);
        Mat warpMat = Imgproc.getPerspectiveTransform(srcPts, dest);
        Mat warped = new Mat();
        Imgproc.warpPerspective(gray, warped, warpMat, new Size(side, side));
        return warped;
    }

    private Point[] orderPoints(Point[] pts) {
        Point[] ordered = new Point[4];
        Arrays.sort(pts, Comparator.comparingDouble(p -> p.x + p.y));
        ordered[0] = pts[0]; // top-left
        ordered[2] = pts[3]; // bottom-right

        Arrays.sort(pts, Comparator.comparingDouble(p -> p.y - p.x));
        ordered[1] = pts[0]; // top-right
        ordered[3] = pts[3]; // bottom-left

        return ordered;
    }
}

File: python/image_preprocessor.py
"""Image preprocessing: detect largest 4-corner contour and warp to square."""
import numpy as np

def preprocess_sudoku_image_from_array(img):
    import cv2

    gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
    gray = cv2.equalizeHist(gray)
    gray = cv2.GaussianBlur(gray, (7, 7), 0)

    thresh = cv2.adaptiveThreshold(gray, 255, cv2.ADAPTIVE_THRESH_GAUSSIAN_C, cv2.THRESH_BINARY_INV, 11, 2)
    kernel = cv2.getStructuringElement(cv2.MORPH_RECT, (3, 3))
    thresh = cv2.morphologyEx(thresh, cv2.MORPH_OPEN, kernel)

    contours, _ = cv2.findContours(thresh, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
    max_area = 0
    max_contour = None
    for cnt in contours:
        area = cv2.contourArea(cnt)
        if area > max_area:
            peri = cv2.arcLength(cnt, True)
            approx = cv2.approxPolyDP(cnt, 0.02 * peri, True)
            if len(approx) == 4:
                max_area = area
                max_contour = approx
    if max_contour is None:
        raise RuntimeError("Could not find Sudoku grid")

    pts = max_contour.reshape(4, 2)
    ordered = order_points(pts)
    side = 450
    dst = np.array([[0, 0], [side - 1, 0], [side - 1, side - 1], [0, side - 1]], dtype="float32")
    M = cv2.getPerspectiveTransform(ordered.astype('float32'), dst)
    warped = cv2.warpPerspective(gray, M, (side, side))
    return warped


def order_points(pts):
    rect = np.zeros((4, 2), dtype="float32")
    s = pts.sum(axis=1)
    rect[0] = pts[np.argmin(s)]
    rect[2] = pts[np.argmax(s)]
    diff = np.diff(pts, axis=1)
    rect[1] = pts[np.argmin(diff)]
    rect[3] = pts[np.argmax(diff)]
    return rect
